ocaml/xcp-rrdd/bin/rrdd/rrdd_server.ml

```diff
@@ -461,36 +461,51 @@ let query_host_ds (ds_name : string) : float =
     Especially, nan, infinity and neg_infinity will be converted to strings 
     "NaN", "infinity" and "-infinity", the client needs to handle by itself.
     *)
-let dump_host_dss_to_file (file : string) : unit =
-  let convert_value x =
-    match classify_float x with
-    | FP_nan ->
-        `String "NaN"
-    | FP_infinite ->
-        `String (if x > 0.0 then "infinity" else "-infinity")
-    | _ ->
-        `Float x
-  in
-  let json =
-    with_lock mutex (fun () ->
-        match !host_rrd with
-        | None ->
-            `Assoc []
-        | Some rrdi ->
-            `Assoc
-              (Rrd.ds_names rrdi.rrd
-              |> List.map (fun ds_name ->
-                     (ds_name, convert_value (query ds_name rrdi))
-                 )
-              )
-    )
-  in
+let convert_value x =
+  match classify_float x with
+  | FP_nan ->
+      `String "NaN"
+  | FP_infinite ->
+      `String (if x > 0.0 then "infinity" else "-infinity")
+  | _ ->
+      `Float x
+
+let generate_json rrd =
+  with_lock mutex (fun () ->
+      match rrd with
+      | None ->
+          `Assoc []
+      | Some rrdi ->
+          `Assoc
+            (Rrd.ds_names rrdi.rrd
+            |> List.map (fun ds_name ->
+                   (ds_name, convert_value (query ds_name rrdi))
+               )
+            )
+  )
+
+let write_json_to_file file json =
   Xapi_stdext_unix.Unixext.atomic_write_to_file file 0o644 (fun fd ->
       let oc = Unix.out_channel_of_descr fd in
       Yojson.Basic.to_channel ~std:true oc json ;
       flush oc
   )
 
+let dump_host_dss_to_file (file : string) : unit =
+  let json = generate_json !host_rrd in
+  write_json_to_file file json
+
+let dump_vm_dss_to_file (file : string) : unit =
+  let json =
```

Comment:
vm_rrds is not protected by a mutex, I think the lock needs to be moved to the outer function to avoid race issues

Reply:
Fixed